// piral-cli-vite/lib/vite/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommonConfig = void 0;
const vite_plugin_codegen_1 = __importDefault(require("vite-plugin-codegen"));
const vite_plugin_environment_1 = __importDefault(require("vite-plugin-environment"));
function createCommonConfig(root, outDir, develop = true, sourcemap = true, minify = true, variables = {}) {
    const varKeys = Object.keys({
        BUILD_PCKG_NAME: process.env.BUILD_PCKG_NAME,
        BUILD_PCKG_VERSION: process.env.BUILD_PCKG_VERSION,
        BUILD_TIME_FULL: process.env.BUILD_TIME_FULL,
        PIRAL_CLI_VERSION: process.env.PIRAL_CLI_VERSION,
        NODE_ENV: process.env.NODE_ENV,
        ...variables,
    });
    return {
        root,
        build: {
            outDir,
            sourcemap,
            emptyOutDir: false,
            minify,
            rollupOptions: {
                output: {
                    assetFileNames: '[name].[hash][extname]',
                    chunkFileNames: '[name].[hash].js',
                    entryFileNames: '[name].[hash].js',
                },
            },
        },
        plugins: [(0, vite_plugin_environment_1.default)(varKeys), (0, vite_plugin_codegen_1.default)({ outDir, rootDir: root })],
    };
}
exports.createCommonConfig = createCommonConfig;
