// vite-plugin-pilet/lib/cjs/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => pilet
});
module.exports = __toCommonJS(src_exports);
var import_magic_string = __toESM(require("magic-string"), 1);
var import_sourcemap_codec = require("@jridgewell/sourcemap-codec");

// src/banner.ts
function modifyImports(ms, dependencies) {
  ms.replace(/\[.*?\]/, (s) => {
    dependencies.forEach((dep) => {
      const depRef = dep.requireId || dep.id;
      s = s.replace(`'${dep.name}'`, `'${depRef}'`);
    });
    return s;
  });
}
function prependBanner(ms, requireRef, dependencies, schema) {
  const deps = dependencies.reduce((deps2, dep) => {
    deps2[dep.id] = dep.ref;
    return deps2;
  }, {});
  if (schema === "v2") {
    ms.prepend(`//@pilet v:2(${requireRef},${JSON.stringify(deps)})
`);
  } else if (schema === "v3") {
    ms.prepend(`//@pilet v:3(${requireRef},${JSON.stringify(deps)})
`);
  }
}
function insertStylesheet(ms, name, debug, schema) {
  if (schema === "v2") {
    const bundleUrl = `function(){try{throw new Error}catch(t){const e=(""+t.stack).match(/(https?|file|ftp|chrome-extension|moz-extension):\\/\\/[^)\\n]+/g);if(e)return e[0].replace(/^((?:https?|file|ftp|chrome-extension|moz-extension):\\/\\/.+)\\/[^\\/]+$/,"$1")+"/"}return"/"}`;
    const cssFiles = ["style.css"];
    const stylesheet = [
      `var d=document`,
      `var __bundleUrl__=(${bundleUrl})()`,
      `${JSON.stringify(cssFiles)}.forEach(cf=>{`,
      `  var u=__bundleUrl__+cf`,
      `  var e=d.createElement("link")`,
      `  e.setAttribute('data-origin', ${JSON.stringify(name)})`,
      `  e.type="text/css"`,
      `  e.rel="stylesheet"`,
      `  e.href=${debug ? 'u+"?_="+Math.random()' : "u"}`,
      `  d.head.appendChild(e)`,
      `})`
    ].join(";\n  ");
    const insertLink = `(function(){
  ${stylesheet};
})()`;
    const execute = "execute: (function () {";
    ms.replace(execute, `${execute}
${insertLink}`);
  } else if (schema === "v3") {
    const cssFiles = ["style.css"];
    const execute = "execute: (function () {";
    const insertLink = `exports("styles", ${JSON.stringify(cssFiles)})`;
    ms.replace(execute, `${execute}
${insertLink}`);
  }
}

// src/index.ts
function pilet({ id, debug, piletName, importmap, requireRef, schema }) {
  const cssFiles = [];
  return {
    name: "pilet",
    transform(_, id2) {
      if (id2.endsWith(".css") || id2.endsWith(".scss") || id2.endsWith(".sass")) {
        cssFiles.push(id2);
      }
    },
    generateBundle(_, bundle) {
      Object.keys(bundle).forEach((file) => {
        const asset = bundle[file];
        if (asset.type === "chunk" && asset.isEntry && asset.name === id) {
          const sm = bundle[`${file}.map`];
          const ms = new import_magic_string.default(asset.code);
          prependBanner(ms, requireRef, importmap, schema);
          asset.code = ms.toString();
          if (sm && "source" in sm && typeof sm.source === "string") {
            const map = JSON.parse(sm.source);
            const arr = (0, import_sourcemap_codec.decode)(map.mappings);
            arr.unshift([]);
            map.mappings = (0, import_sourcemap_codec.encode)(arr);
            sm.source = JSON.stringify(map);
          }
        }
      });
    },
    renderChunk(content, asset) {
      const ms = new import_magic_string.default(content);
      modifyImports(ms, importmap);
      if (asset.isEntry && asset.name === id && cssFiles.length) {
        insertStylesheet(ms, piletName, debug, schema);
      }
      return {
        code: ms.toString(),
        map: ms.generateMap({ hires: true })
      };
    }
  };
}
